src/libAtomVM/otp_socket.c

```diff
@@ -0,0 +1,1364 @@
+
+    UNUSED(argc);
+
+    GlobalContext *global = ctx->global;
+
+    term domain_term = argv[0];
+    if (term_is_invalid_term(domain_term)) {
+        domain_term = globalcontext_make_atom(global, inet_atom);
+    }
+    term type_term = argv[1];
+    if (term_is_invalid_term(type_term)) {
+        type_term = globalcontext_make_atom(global, stream_atom);
+    }
+    term protocol_term = argv[2];
+    if (term_is_invalid_term(protocol_term)) {
+        protocol_term = globalcontext_make_atom(global, tcp_atom);
+    }
+
+    int fd = socket(get_domain(global, domain_term), get_type(global, type_term), get_protocol(global, protocol_term));
+    if (fd == -1) {
+
+        AVM_LOGE(TAG, "Failed to initialize socket.");
+
+        if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+            AVM_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+
+        term error_tuple = term_alloc_tuple(2, &ctx->heap);
+        term_put_tuple_element(error_tuple, 0, ERROR_ATOM);
+        term_put_tuple_element(error_tuple, 1, posix_errno_to_term(errno, global));
+        return error_tuple;
+
+    } else {
+
+        struct SocketFd *fd_obj = enif_alloc_resource(socket_resource_type, sizeof(struct SocketFd));
+
+        if (IS_NULL_PTR(fd_obj)) {
+            AVM_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+
+        fd_obj->fd = fd;
+        fd_obj->selecting_process_id = INVALID_PROCESS_ID;
+        fd_obj->pid = term_invalid_term();
+        TRACE("nif_socket_open: Created socket fd=%i\n", fd_obj->fd);
+
+        term obj = enif_make_resource(erl_nif_env_from_context(ctx), fd_obj);
+        enif_release_resource(fd_obj);
+
+        size_t requested_size = TUPLE_SIZE(2) + TUPLE_SIZE(2) + REF_SIZE;
+        if (UNLIKELY(memory_ensure_free_with_roots(ctx, requested_size, 1, &obj, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+            AVM_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+
+        term socket_term = term_alloc_tuple(2, &ctx->heap);
+        uint64_t ref_ticks = globalcontext_get_ref_ticks(ctx->global);
+        term ref = term_from_ref_ticks(ref_ticks, &ctx->heap);
+        term_put_tuple_element(socket_term, 0, obj);
+        term_put_tuple_element(socket_term, 1, ref);
+
+        term result = term_alloc_tuple(2, &ctx->heap);
+        term_put_tuple_element(result, 0, OK_ATOM);
+        term_put_tuple_element(result, 1, socket_term);
+
+        return result;
+    }
+}
+
+static term get_socket(term socket_term)
+{
+    return term_get_tuple_element(socket_term, 0);
+}
+
+static bool term_is_socket(term socket_term)
+{
+    bool ret = term_is_tuple(socket_term)
+        && term_get_tuple_arity(socket_term) == 2
+        && term_is_binary(term_get_tuple_element(socket_term, 0))
```

Comment:
We probably want to implement `term_is_resource` for the day we'll move to magic references. Alternatively, you could skip this check, enif_get_resource is fail proof.

Reply:
I think I will leave it here for completeness/readability/following existing patterns.  Is that okay?

Reply:
Sure. As tests will break if resources are made magic references, we will write a `term_is_resource` function then.